pkg/skaffold/build/scheduler.go

```diff
@@ -35,88 +35,85 @@ type ArtifactBuilder func(ctx context.Context, out io.Writer, artifact *latest.A
-	runInSequence = InSequence
+	buffSize = bufferedLinesPerArtifact
 )
 
-// InParallel builds a list of artifacts in parallel but prints the logs in sequential order.
-func InParallel(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact, buildArtifact ArtifactBuilder, concurrency int) ([]Artifact, error) {
-	if len(artifacts) == 0 {
-		return nil, nil
-	}
-
-	if len(artifacts) == 1 || concurrency == 1 {
-		return runInSequence(ctx, out, tags, artifacts, buildArtifact)
-	}
+type scheduler struct {
+	artifacts       []*latest.Artifact
+	artifactBuilder ArtifactBuilder
 
-	var wg sync.WaitGroup
-	defer wg.Wait()
+	nodes []buildNode // size len(artifacts)
 
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
+	outputs []chan string // size len(artifacts)
+	results sync.Map      // map[artifact.name]error
 
-	results := new(sync.Map)
-	outputs := make([]chan string, len(artifacts))
+	sem countingSemaphore
+	wg  sync.WaitGroup
+}
 
-	if concurrency == 0 {
-		concurrency = len(artifacts)
+func newScheduler(artifacts []*latest.Artifact, artifactBuilder ArtifactBuilder, concurrency int) *scheduler {
+	s := scheduler{
+		artifacts:       artifacts,
+		artifactBuilder: artifactBuilder,
+		nodes:           createNodes(artifacts),
+		outputs:         make([]chan string, len(artifacts)),
+		sem:             newCountingSemaphore(concurrency),
 	}
-	sem := make(chan bool, concurrency)
-
-	// Run builds in //
-	wg.Add(len(artifacts))
+	s.wg.Add(len(artifacts))
 	for i := range artifacts {
-		outputs[i] = make(chan string, buffSize)
-		r, w := io.Pipe()
+		s.outputs[i] = make(chan string, buffSize)
+	}
+	return &s
+}
 
-		// Run build and write output/logs to piped writer and store build result in
-		// sync.Map
-		go func(i int) {
-			sem <- true
-			runBuild(ctx, w, tags, artifacts[i], results, buildArtifact)
-			<-sem
+func (s *scheduler) run(ctx context.Context, out io.Writer, tags tag.ImageTags) ([]Artifact, error) {
+	for i := range s.artifacts {
+		r, w := io.Pipe()
 
-			wg.Done()
-		}(i)
+		// Create a goroutine for each element in dag. Each goroutine waits on its dependencies to finish building.
+		// Because our artifacts form a DAG, at least one of the goroutines should be able to start building.
+		go s.build(ctx, w, tags, i)
 
 		// Read build output/logs and write to buffered channel
-		go readOutputAndWriteToChannel(r, outputs[i])
+		go readOutputAndWriteToChannel(r, s.outputs[i])
 	}
+	s.wg.Wait()
 
 	// Print logs and collect results in order.
-	return collectResults(out, artifacts, results, outputs)
+	return collectResults(out, s.artifacts, &s.results, s.outputs)
 }
 
-func runBuild(ctx context.Context, cw io.WriteCloser, tags tag.ImageTags, artifact *latest.Artifact, results *sync.Map, build ArtifactBuilder) {
-	event.BuildInProgress(artifact.ImageName)
+func (s *scheduler) build(ctx context.Context, cw io.WriteCloser, tags tag.ImageTags, i int) {
+	defer cw.Close()
+	defer s.wg.Done()
```

Comment:
`defer` calls go onto a function-specific stack, so this will call `s.wg.Done()` before `cw.Close()`. I think they should be reversed no?

Reply:
yes, you're right. Fixed.